modules/assim.sequential/R/sda.enkf.R

```diff
@@ -129,176 +145,382 @@ sda.enkf <- function(settings,IC,prior,obs){
+        
+        ## update parameters  
+        dat = as.matrix(jdat)
+        dat = dat[3000:10000,]
+        iq = grep("q",colnames(dat))
+        iX = grep("X[",colnames(dat),fixed=TRUE)
+        mu.a  = colMeans(dat[,iX])
+        Pa  = cov(dat[,iX])
+        Pa[is.na(Pa)]<- 0 
+        
+        CI.X1[,t] = quantile(dat[,iX[1]],c(0.025,0.5,0.975))
+        CI.X2[,t] = quantile(dat[,iX[2]],c(0.025,0.5,0.975))
+        
+        mq = dat[,iq] #Omega, Precision
+        q.bar = matrix(apply(mq,2,mean),length(mu.f),length(mu.f)) #Mean Omega, Precision
+        
+        col = matrix(1:length(mu.f)^2,length(mu.f),length(mu.f))
+        WV = matrix(0,length(mu.f),length(mu.f))
+        for(i in 1:length(mu.f)){
+          for(j in 1:length(mu.f)){
+            WV[i,j] <- wish.df(q.bar, X = mq, i=i, j=j, col=col[i,j])
+          }
+        }
+        
+        n = mean(WV) #n + 1
+        if(n < length(mu.f)) n = length(mu.f)
+        V = solve(q.bar)*n
+        
+        #ifelse(eigen(V)$values>0,eigen(V)$values,print("matrix not positive definite"))
+        
+        aqq[t+1,,] = V
+        bqq[t+1] = n
+    }
+
+    enkf.params[[t]] = list(mu.f = mu.f, Pf = Pf, mu.a = mu.a, Pa = Pa) 
  
  ## update state matrix
-analysis = as.data.frame(rmvnorm(nens,mu.a,Pa,method="svd"))
-names(analysis) = names(X)
- # EAKF
- if(FALSE){
-   analysis = X
-   
-   ## Math from Anderson 2001. gives correct mean but incorrect var
-   A.svd = svd(Pf)
-   F = A.svd$v
-   G = diag(sqrt(A.svd$d))
-   B.svd = svd(t(G)%*%t(F)%*%t(H)%*%solve(R)%*%H%*%F%*%G)
-   U = B.svd$v
-   B = diag((1+B.svd$d)^(-0.5))
-   A = solve(t(F))%*%t(G)*solve(t(U))%*%t(B)%*%solve(t(G))%*%t(F)
-   for(i in 1:nens){
-     analysis[i,] = t(A)%*%matrix(as.numeric(X[i,])-mu.f)+mu.a
-   }
-   
-   ## HACK IGNORNING COVARIANCE
-   for(i in 1:nens){
-     analysis[i,] = mu.a + (matrix(as.numeric(X[i,]))-mu.f)*sqrt(diag(Pa)/diag(Pf))
-   }   
-   
- }
+    analysis = as.data.frame(rmvnorm(nens,mu.a,Pa,method="svd"))
+    #analysis = exp(analysis)
+    analysis[is.na(analysis)] <- 0
+    analysis <- abs(analysis)
+    names(analysis) = names(X)
+#  # EAKF
+#  if(FALSE){
+#    analysis = X
+#    
+#    ## Math from Anderson 2001. gives correct mean but incorrect var
+#    A.svd = svd(Pf)
+#    F = A.svd$v
+#    G = diag(sqrt(A.svd$d))
+#    B.svd = svd(t(G)%*%t(F)%*%t(H)%*%solve(R)%*%H%*%F%*%G)
+#    U = B.svd$v
+#    B = diag((1+B.svd$d)^(-0.5))
+#    A = solve(t(F))%*%t(G)*solve(t(U))%*%t(B)%*%solve(t(G))%*%t(F)
+#    for(i in 1:nens){
+#      analysis[i,] = t(A)%*%matrix(as.numeric(X[i,])-mu.f)+mu.a
+#    }
+#    
+#    ## HACK IGNORNING COVARIANCE
+#    for(i in 1:nens){
+#      analysis[i,] = mu.a + (matrix(as.numeric(X[i,]))-mu.f)*sqrt(diag(Pa)/diag(Pf))
+#    }   
+#    
+#  }
 ## analysis sanity check
-for(i in 2:ncol(analysis)){
-  analysis[analysis[,i]<0,i] = 0.0
-}
+#for(i in 2:ncol(analysis)){
+#  analysis[analysis[,i]<0,i] = 0.0
```

Comment:
Willing to admit that I was previously making the same bad assumption about Analysis truncation that I just criticized @araiho on.